File: dgp.py
import numpy as np
from mcmc import gibbs_sampling
from tooling import sq_dist, krig

def fit_dgp(x, y, nmcmc=10000, D=None, verb=True, w_0=None, g_0=0.01, 
            theta_y_0=0.1, theta_w_0=0.1, true_g=None, cov="matern", v=2.5):
    """Fit Deep Gaussian Process."""
    if D is None:
        D = x.shape[1]
    
    if cov == "exp2":
        v = None
    
    if w_0 is None:
        w_0 = np.tile(x, [1, D])
    
    settings = {
        'l': 1,
        'u': 2,
        'alpha': {'g': 1.5, 'theta_w': 1.5, 'theta_y': 1.5},
        'beta': {'g': 3.9, 'theta_w': 3.9/4, 'theta_y': 3.9/6},
        'inner_tau2': 1
    }
    
    initial = {
        'w': w_0,
        'theta_y': theta_y_0,
        'theta_w': theta_w_0,
        'g': g_0,
        'tau2': 1
    }
    
    if cov == "matern" and v not in [0.5, 1.5, 2.5]:
        raise ValueError('v must be one of 0.5, 1.5 or 2.5')
    
    output = {
        'x': x,
        'y': y,
        'nmcmc': nmcmc,
        'settings': settings,
        'v': v
    }
    
    samples = gibbs_sampling(x, y, nmcmc, D, verb, initial, true_g, settings, v)
    output.update(samples)
    
    return output

def predict_dgp(object, x_new):
    """Make predictions using fitted DGP model."""
    object['x_new'] = x_new
    n_new = x_new.shape[0]
    D = object['w'][0].shape[1]
    dx = sq_dist(object['x'])
    d_cross = sq_dist(x_new, object['x'])
    mu_t = np.full((n_new, object['nmcmc']), np.nan)
    sigma_sum = np.zeros((n_new, n_new))
    
    for t in range(object['nmcmc']):
        w_t = object['w'][t]
        w_new = np.full((n_new, D), np.nan)
        
        for i in range(D):
            k = krig(w_t[:, i], dx, None, d_cross, object['theta_w'][t, i], 
                    np.sqrt(np.finfo(float).eps), object['settings']['inner_tau2'], 
                    None, None, object['v'])
            w_new[:, i] = k['mean']
        
        k = krig(object['y'], sq_dist(w_t), sq_dist(w_new), sq_dist(w_new, w_t), 
                object['theta_y'][t], object['g'][t], object['tau2'][t], 
                False, True, object['v'])
        mu_t[:, t] = k['mean'][:,0]
        sigma_sum += k['sigma']
    
    mu_cov = np.cov(mu_t)
    object['mean'] = np.mean(mu_t, axis=1)
    object['sigma'] = sigma_sum/object['nmcmc'] + mu_cov
    
    return object

File: dgp_cpp/src/dgp.cpp
// top-level fit / predict for the two-layer Deep GP.
#include "dgp.hpp"
#ifdef _OPENMP
#include <omp.h>
#endif

namespace dgp {

DGPModel fit_dgp(const MatrixXd& x, const VectorXd& y, int nmcmc, int D,
                 bool verb, double g_0, double theta_y_0, double theta_w_0,
                 const double* true_g, double v, std::mt19937_64& rng) {
    DGPModel m;
    m.x = x;
    m.y = y;
    m.nmcmc = nmcmc;
    m.D = D;
    m.v = v;
    m.settings = Settings{};

    // w_0 = np.tile(x, [1, D]); for a single-column design this repeats x.
    const int n = static_cast<int>(x.rows());
    const int p = static_cast<int>(x.cols());
    Initial initial;
    initial.w = MatrixXd(n, D);
    for (int i = 0; i < D; ++i) initial.w.col(i) = x.col(i % p);
    initial.theta_y = theta_y_0;
    initial.theta_w = theta_w_0;
    initial.g = g_0;
    initial.tau2 = 1.0;

    m.samples = gibbs_sampling(x, y, nmcmc, D, verb, initial, true_g,
                               m.settings, v, rng);
    return m;
}

void predict_dgp(DGPModel& m, const MatrixXd& x_new) {
    m.x_new = x_new;
    const int n_new = static_cast<int>(x_new.rows());
    const int nmcmc = m.nmcmc;
    const int D = static_cast<int>(m.samples.w[0].cols());
    const double v = m.v;
    const double inner_tau2 = m.settings.inner_tau2;

    const MatrixXd dx = sq_dist(m.x);
    const MatrixXd d_cross = sq_dist(x_new, m.x);   // (n_new x n)

    MatrixXd mu_t(n_new, nmcmc);

#ifdef _OPENMP
    const int nthreads = omp_get_max_threads();
#else
    const int nthreads = 1;
#endif
    std::vector<MatrixXd> sigma_local(nthreads,
                                      MatrixXd::Zero(n_new, n_new));

    #pragma omp parallel for schedule(dynamic)
    for (int t = 0; t < nmcmc; ++t) {
#ifdef _OPENMP
        const int tid = omp_get_thread_num();
#else
        const int tid = 0;
#endif
        const MatrixXd& w_t = m.samples.w[t];

        // layer 1: map the test inputs through each latent dimension.
        MatrixXd w_new(n_new, D);
        for (int i = 0; i < D; ++i) {
            KrigResult k = krig(w_t.col(i), dx, nullptr, &d_cross,
                                m.samples.theta_w(t, i), sqeps(), inner_tau2,
                                false, false, v);
            w_new.col(i) = k.mean;
        }

        // layer 2: GP on the latent layer -> predictive mean & covariance.
        MatrixXd dwt = sq_dist(w_t);
        MatrixXd dnew = sq_dist(w_new);
        MatrixXd dcr = sq_dist(w_new, w_t);
        KrigResult k = krig(m.y, dwt, &dnew, &dcr,
                            m.samples.theta_y(t), m.samples.g(t),
                            m.samples.tau2(t), false, true, v);
        mu_t.col(t) = k.mean;
        sigma_local[tid] += k.sigma;
    }

    MatrixXd sigma_sum = MatrixXd::Zero(n_new, n_new);
    for (int th = 0; th < nthreads; ++th) sigma_sum += sigma_local[th];

    // posterior predictive mean and (law-of-total-variance) covariance.
    m.mean = mu_t.rowwise().mean();
    MatrixXd centered = mu_t.colwise() - m.mean;            // (n_new x nmcmc)
    MatrixXd mu_cov = (centered * centered.transpose()) / (nmcmc - 1.0);
    m.sigma = sigma_sum / static_cast<double>(nmcmc) + mu_cov;
    m.has_prediction = true;
}

} // namespace dgp
